rb/lib/selenium/webdriver/remote/response.rb

```diff
@@ -59,34 +56,12 @@ def assert_ok
           raise Error::ServerError, self
         end
 
-        def add_backtrace(ex, server_trace)
-          return unless server_trace
-
-          backtrace = case server_trace
-                      when Array
-                        backtrace_from_remote(server_trace)
-                      when String
-                        server_trace.split("\n")
-                      end
-
-          ex.set_backtrace(backtrace + ex.backtrace)
-        end
-
-        def backtrace_from_remote(server_trace)
-          server_trace.filter_map do |frame|
-            next unless frame.is_a?(Hash)
-
-            file = frame['fileName']
-            line = frame['lineNumber']
-            meth = frame['methodName']
-
-            class_name = frame['className']
-            file = "#{class_name}(#{file})" if class_name
-
-            meth = 'unknown' if meth.nil? || meth.empty?
-
-            "[remote server] #{file}:#{line}:in `#{meth}'"
-          end
+        def add_cause(ex, error, backtrace)
+          raise Error::WebDriverError, backtrace
+        rescue Error::WebDriverError
+          raise ex
+        rescue Error.for_error(error)
+          ex
```

Comment:
I think throwing the execption just to set the backtrace is a bit inefficient/unclear:

```suggestion
          cause = Error::WebDriverError.new
          cause.set_backtrace(backtrace)
          raise ex, cause: cause
        rescue Error.for_error(error)
          ex
```

Reply:
@ioquatix I can totally see what you are saying, if I tried the suggested implementation we get the cause without the backtrace included:

This is the response:

```Selenium::WebDriver::Error::WebDriverError ``

However if we do:

e.cause.backtrace, we get this:

```Ruby
["0   chromedriver                        0x0000000100cc6a80 chromedriver + 4385408\n" +
 "1   chromedriver                        0x0000000100cbf38c chromedriver + 4354956\n" +
 "2   chromedriver                        0x00000001008dcb0c chromedriver + 281356\n" +
 "3   chromedriver                        0x000000010091f2f8 chromedriver + 553720\n" +
 "4   chromedriver                        0x0000000100957d24 chromedriver + 785700\n" +
 "5   chromedriver                        0x0000000100913eec chromedriver + 507628\n" +
 "6   chromedriver                        0x00000001009148c4 chromedriver + 510148\n" +
 "7   chromedriver                        0x0000000100c8e43c chromedriver + 4154428\n" +
 "8   chromedriver                        0x0000000100c92ea0 chromedriver + 4173472\n" +
 "9   chromedriver                        0x0000000100c73ff8 chromedriver + 4046840\n" +
 "10  chromedriver                        0x0000000100c9378c chromedriver + 4175756\n" +
 "11  chromedriver                        0x0000000100c66fb8 chromedriver + 3993528\n" +
 "12  chromedriver                        0x0000000100cb121c chromedriver + 4297244\n" +
 "13  chromedriver                        0x0000000100cb1398 chromedriver + 4297624\n" +
 "14  chromedriver                        0x0000000100cbef84 chromedriver + 4353924\n" +
 "15  libsystem_pthread.dylib             0x00000001940cef94 _pthread_start + 136\n" +
 "16  libsystem_pthread.dylib             0x00000001940c9d34 thread_start + 8\n"]
 ```
 
 On the previous implementation cause returns:
 
 ```Ruby
 #<Selenium::WebDriver::Error::WebDriverError: 0   chromedriver                        0x0000000100ac6a80 chromedriver + 4385408
1   chromedriver                        0x0000000100abf38c chromedriver + 4354956
2   chromedriver                        0x00000001006dcb0c chromedriver + 281356
3   chromedriver                        0x000000010071f2f8 chromedriver + 553720
4   chromedriver                        0x0000000100757d24 chromedriver + 785700
5   chromedriver                        0x0000000100713eec chromedriver + 507628
6   chromedriver                        0x00000001007148c4 chromedriver + 510148
7   chromedriver                        0x0000000100a8e43c chromedriver + 4154428
8   chromedriver                        0x0000000100a92ea0 chromedriver + 4173472
9   chromedriver                        0x0000000100a73ff8 chromedriver + 4046840
10  chromedriver                        0x0000000100a9378c chromedriver + 4175756
11  chromedriver                        0x0000000100a66fb8 chromedriver + 3993528
12  chromedriver                        0x0000000100ab121c chromedriver + 4297244
13  chromedriver                        0x0000000100ab1398 chromedriver + 4297624
14  chromedriver                        0x0000000100abef84 chromedriver + 4353924
15  libsystem_pthread.dylib             0x00000001940cef94 _pthread_start + 136
16  libsystem_pthread.dylib             0x00000001940c9d34 thread_start + 8
>
```

Because the backtrace is passed to the cause, and e.backtrace.cause returns:

```Ruby
["/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/response.rb:60:in `add_cause'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/response.rb:41:in `error'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/response.rb:52:in `assert_ok'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/response.rb:34:in `initialize'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/http/common.rb:101:in `new'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/http/common.rb:101:in `create_response'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/http/default.rb:103:in `request'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/http/common.rb:67:in `call'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/bridge.rb:675:in `execute'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/remote/bridge.rb:543:in `find_element_by'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/lib/selenium/webdriver/common/search_context.rb:71:in `find_element'",
 "/Users/apeque01/Desktop/main_folder/Projects/open_source/selenium/rb/spec/integration/selenium/webdriver/error_spec.rb:40:in `block (2 levels) in <module:WebDriver>'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:263:in `instance_exec'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:263:in `block in run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:511:in `block in with_around_and_singleton_context_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:468:in `block in with_around_example_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:486:in `block in run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:626:in `block in run_around_example_hooks_for'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:352:in `call'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-expectations-3.13.1/lib/rspec/expectations/failure_aggregator.rb:25:in `block in aggregate'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-support-3.13.1/lib/rspec/support.rb:126:in `with_failure_notifier'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-expectations-3.13.1/lib/rspec/expectations/failure_aggregator.rb:23:in `aggregate'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-expectations-3.13.1/lib/rspec/matchers.rb:306:in `aggregate_failures'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/configuration.rb:2281:in `block in define_built_in_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:457:in `instance_exec'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:457:in `instance_exec'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:390:in `execute_with'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:628:in `block (2 levels) in run_around_example_hooks_for'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:352:in `call'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:629:in `run_around_example_hooks_for'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/hooks.rb:486:in `run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:468:in `with_around_example_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:511:in `with_around_and_singleton_context_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example.rb:259:in `run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:646:in `block in run_examples'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:642:in `map'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:642:in `run_examples'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/example_group.rb:607:in `run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:121:in `block (3 levels) in run_specs'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:121:in `map'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:121:in `block (2 levels) in run_specs'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/configuration.rb:2091:in `with_suite_hooks'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:116:in `block in run_specs'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/reporter.rb:74:in `report'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:115:in `run_specs'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:89:in `run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:71:in `run'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/lib/rspec/core/runner.rb:45:in `invoke'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/gems/rspec-core-3.13.0/exe/rspec:4:in `<top (required)>'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/bin/rspec:25:in `load'",
 "/Users/apeque01/.rvm/gems/ruby-3.0.0/bin/rspec:25:in `<main>'"]
 ```
 
 I like your suggested implementation way more, and I think is clearer, but I just want to check with @p0deje and @titusfortner what is our preference regarding the content of cause

Reply:
I would have expected these to be identical - is cause not an array of strings?

Reply:
Yes exactly

Reply:
Previously it was being split:

```ruby
          backtrace = case server_trace
                      when Array
                        backtrace_from_remote(server_trace)
                      when String
                        server_trace.split("\n")
                      end
```

Is that still the case?

Reply:
So actually I just re-test it and if we want to keep the backtrace in the cause, we have to pass the backtrace to the creation of the error so it will look like

```Ruby
        def add_cause(ex, error, backtrace)
          cause = Error::WebDriverError.new(backtrace)
          cause.set_backtrace(backtrace)
          raise ex, cause: cause
        rescue Error.for_error(error)
          ex
        end
```

I just made a new commit including this :)